### PR TITLE
Admin ticket reservations and user registration

### LINKED ISSUE #602: Work out a clean way for people to pay for tickets outside the usual flow
As usual, we need to let people buy tickets outside of the usual rounds, and without forcing a payment type on them.

Previously we just created a new ticket type and added a token, but that now involves a ProductViewProduct (and new ProductGroup if we want to limit them).

We could make this easier, or create a way to pre-populate a basket and start them at the /ticket/pay step. Alternatively, allow payments that are "new" to change provider.

### PR DESCRIPTION
**Pull request by marksteward**
Allow admins to:
 - give a user a link to let them create an account (/admin/user/signup)
 - reserve tickets for a user once they have an account (/admin/tickets/reserve)

Allow users to:
 - sign up with an admin-provided link (/signup)
 - recover their reserved tickets into their basket (/tickets/reserved)

Also tidy the login and "my account" pages and fix allow_promo.

Security: this changes verify_login_code into a slightly more generic verify_hmac_msg.

Fixes #602, #438

### CODE REVIEW

File: apps/users.py [hunk 1]
```diff
@@ -106,6 +111,71 @@ def logout():
...
+    name = StringField('Name', [Required()])
+    allow_promo = BooleanField('Send me occasional emails about future EMF events')
+    signup = SubmitField('Sign up')
+
+    def validate_email(form, field):
+        if User.does_user_exist(field.data):
+            field.was_duplicate = True
+
+            msg = Markup(
+                render_template_string('Account already exists. '
```
Comment (SamLR) [inline apps/users.py]: will these strings be auto-concatenated?

File: apps/users.py [hunk 2]
```diff
@@ -106,6 +111,71 @@ def logout():
...
+
+@users.route("/signup", methods=['GET', 'POST'])
+def signup():
+    if not request.args.get('code'):
+        abort(404)
+
+    uid = verify_signup_code(app.config['SECRET_KEY'], time.time(), request.args.get('code'))
+    if uid is None:
+        flash("Your signup link was invalid. Please note that they expire after 6 hours.")
+        return redirect(url_for('.account'))
```
Comment (SamLR) [inline apps/users.py]: Shouldn't this redirect to a login page rather than the account page?

File: apps/users.py (hunk 1 above)
Comment (marksteward) [reply apps/users.py]: yep

File: apps/users.py (hunk 2 above)
Comment (marksteward) [reply apps/users.py]: it will, indirectly. I'll make that more explicit

**Pull request merged**